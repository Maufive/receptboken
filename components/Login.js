import React, { Component } from "react";
import axios from "axios";
import Link from "next/link";
import { LoginForm } from "./styles/ModalStyles";
import { Button } from "./styles/Button";
import LockIcon from "../svg/lock.svg";
import MailIcon from "../svg/mail.svg";

class Login extends Component {
	constructor() {
		super();
		this.state = {
			email: "", // hej@gmail.com .. test6@gmail.com ..
			password: "", //asdf
			error: null
		};
	}

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	onSubmit = e => {
		e.preventDefault();
		const { email, password } = this.state;
		const { setUser, setMessage } = this.props;
		axios
			.post(`https://receptboken.herokuapp.com/auth/login`, { email, password })
			.then(result => {
				localStorage.setItem("jwtToken", result.data.token);
				setUser(result.data.user);
				setMessage("success", `Välkommen ${result.data.user.fname}!`);
				// Router.push("/");
			})
			.catch(error => {
				console.log(error.response.data.message);
				this.setState({ error: error.response.data.message });
			});
	};

	render() {
		return (
			<div style={{ margin: "0 auto" }}>
				<LoginForm onSubmit={this.onSubmit}>
					{this.state.error ? (
						<h4 style={{ color: "#bc1616", marginBottom: "2.5rem" }}>
							{this.state.error}
						</h4>
					) : (
						<h2>Vänligen logga in</h2>
					)}
					<div>
						<input
							type="email"
							name="email"
							id="email"
							value={this.state.email}
							onChange={this.saveToState}
							required
						/>
						<label htmlFor="email">
							<MailIcon /> Email
						</label>
					</div>
					<div>
						<input
							type="password"
							name="password"
							id="password"
							value={this.state.password}
							onChange={this.saveToState}
							required
						/>
						<label htmlFor="password">
							<LockIcon /> Lösenord
						</label>
					</div>
					<Button fullWidth type="submit" id="login-button">
						Logga in →
					</Button>
					<h3 style={{ color: "#5A5555" }}>Inget konto?</h3>
					<Link href="/register">
						<a>
							<Button
								fullWidth
								primary
								id="register-button"
								onClick={this.props.closeModal}
							>
								Registrera →
							</Button>
						</a>
					</Link>
				</LoginForm>
			</div>
		);
	}
}

export default Login;
