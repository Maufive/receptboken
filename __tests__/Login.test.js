import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Login from "../components/Login";

describe("<Login />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<Login />)));

	it("renders to the page", () => {
		expect(wrapper.find("ModalStyles__LoginForm").exists()).toBe(true);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	it("renders an email input", () => {
		expect(wrapper.find("#email").length).toEqual(1);
	});

	it("render an password input", () => {
		expect(wrapper.find("#password").length).toEqual(1);
	});

	it("renders the login-button", () => {
		expect(wrapper.find("#login-button").length).toEqual(1);
	});

	it("renders the register-button", () => {
		expect(wrapper.find("#register-button").length).toEqual(1);
	});

	describe("Email input", () => {
		it("should respond to change event and change the state of the Login-component", () => {
			wrapper.find("#email").simulate("change", {
				target: { name: "email", value: "test@gmail.com" }
			});

			expect(wrapper.state("email")).toEqual("test@gmail.com");
		});
	});

	describe("Password input", () => {
		it("should respond to change event and change the state of the Login-component", () => {
			wrapper.find("#password").simulate("change", {
				target: { name: "password", value: "password123" }
			});

			expect(wrapper.state("password")).toEqual("password123");
		});
	});

	describe("Submit the loginform", () => {
		it("throws error without input", () => {
			const mockFn = {
				preventDefault: jest.fn()
			};
			expect(() =>
				wrapper.find("ModalStyles__LoginForm").simulate("submit", mockFn)
			).toThrow();
			expect(mockFn.preventDefault).toHaveBeenCalledTimes(1);
		});

		it("does not throw error with input from the user", () => {
			const mockFn = {
				preventDefault: jest.fn()
			};
			wrapper.setState({
				email: "niklas@gmail.com",
				password: "secretpassword123"
			});
			wrapper.update();
			wrapper.find("ModalStyles__LoginForm").simulate("submit", mockFn);
			expect(mockFn.preventDefault).toHaveBeenCalledTimes(1);
		});
	});
});
