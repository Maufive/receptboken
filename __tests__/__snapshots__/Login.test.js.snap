// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders to the page 1`] = `
<div
  style={
    Object {
      "margin": "0 auto",
    }
  }
>
  <ModalStyles__LoginForm
    onSubmit={[Function]}
  >
    <h2>
      Vänligen logga in
    </h2>
    <div>
      <input
        id="email"
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
        value=""
      />
      <label
        htmlFor="email"
      >
        <MailIcon
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        />
         Email
      </label>
    </div>
    <div>
      <input
        id="password"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
      <label
        htmlFor="password"
      >
        <LockIcon
          viewBox="0 0 58 58"
          xmlns="http://www.w3.org/2000/svg"
        />
         Lösenord
      </label>
    </div>
    <Button
      fullWidth={true}
      id="login-button"
      type="submit"
    >
      Logga in →
    </Button>
    <h3
      style={
        Object {
          "color": "#5A5555",
        }
      }
    >
      Inget konto?
    </h3>
    <Link
      href="/register"
    >
      <a>
        <Button
          fullWidth={true}
          id="register-button"
          primary={true}
        >
          Registrera →
        </Button>
      </a>
    </Link>
  </ModalStyles__LoginForm>
</div>
`;
